### PR TITLE
add safety checker

### PR DESCRIPTION
**Pull request by xijia-tao**
- [x] add safety checker in `stable_diffusion_v2/tools/safety_checker`
- [x] support both mindspore 1.x and 2.0
- [x] move clip implementation from `stable_diffusion_v2/eval/clip_score/clip` to `stable_diffusion_v2/ldm/models/clip`
- [x] rename `load_checkpoint` arg to `ckpt_path`
- [x] update related documentation in in `stable_diffusion_v2/docs/en/safety_check.md`
- [x] use safety checker in `text_to_image.py`, enable the check with `--check_safety`

### CODE REVIEW

File: examples/stable_diffusion_v2/text_to_image.py [hunk 1]
```diff
@@ -183,6 +184,15 @@ def main(args):
         sampler = PLMSSampler(model)
         sname = "plms"
 
+    # create safety checker
+    if args.check_safety:
+        if args.safety_version == "1":
+            safety_checker = SafetyChecker1(backend="ms", ckpt_path=args.clip_ckpt_path)
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/text_to_image.py:190]: Rename SafetyChecker1 -> SafetyCheckerV1
SafetyChecker2-> SafetyCheckerV2.

Similarly for the file name.

File: examples/stable_diffusion_v2/docs/en/safety_check.md
Comment (SamitHuang) [inline examples/stable_diffusion_v2/docs/en/safety_check.md:1]: Need better structuring.

# Safety Checker

## Introduction
...
## Usage
...

File: examples/stable_diffusion_v2/tools/eval/eval_clip_i_score.py [hunk 2]
```diff
@@ -64,18 +64,18 @@ def load_images(image_path_or_dir):
...
-        default="bpe_simple_vocab_16e6.txt.gz",
+        default="ldm/models/clip/bpe_simple_vocab_16e6.txt.gz",
         type=str,
-        help="load model checkpoint." " Default: bpe_simple_vocab_16e6.txt.gz",
+        help="load model checkpoint." " Default: ldm/models/clip/bpe_simple_vocab_16e6.txt.gz",
     )
     parser.add_argument("--quiet", action="store_true", help="set this flag to avoid printing scores")
     parser.add_argument(
-        "--no_check_certificate",
+        "--check_certificate",
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/tools/eval/eval_clip_i_score.py:76]: check_certificate is a too trivial arg, we can leave it to False always in this case.

File: examples/stable_diffusion_v2/tools/eval/eval_clip_score.py [hunk 3]
```diff
@@ -64,9 +68,9 @@
     )
     parser.add_argument("--quiet", action="store_true", help="set this flag to avoid printing scores")
     parser.add_argument(
-        "--no_check_certificate",
+        "--check_certificate",
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/tools/eval/eval_clip_score.py:71]: rm this arg.

File: examples/stable_diffusion_v2/tools/safety_checker/l14_nsfw.ckpt
Comment (SamitHuang) [inline examples/stable_diffusion_v2/tools/safety_checker/l14_nsfw.ckpt:1]: We should put the checkpoint on download server instead.

File: examples/stable_diffusion_v2/tools/safety_checker/safety_checker_1.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        help="backend to do CLIP model inference for CLIP score compute. Option: ms, pt." " Default: ms",
+    )
+    parser.add_argument(
+        "--tokenizer_path",
+        default="ldm/models/clip/bpe_simple_vocab_16e6.txt.gz",
+        type=str,
+        help="load tokenizer checkpoint." " Default: ldm/models/clip/bpe_simple_vocab_16e6.txt.gz",
+    )
+    parser.add_argument(
+        "--settings_path",
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/tools/safety_checker/safety_checker_1.py:207]: For simplicity, let us just use one argument to parse config file. In `safety_setting_1.yaml` (please rename to `safety_checker_v1.yaml`),  indicate the path to the config file of clip model.

File: examples/stable_diffusion_v2/docs/en/safety_check.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+You can use the checkers in your image generation pipelines. It is by default included in `text_to_image.py` for checking the safety of stable diffusion outputs. In addition, you can test the checkers on saved images. In `examples/stable_diffusion` directory, please run,
+
+- MindSpore backend
+```
+python tools/safety_checker/safety_checker_1.py --ckpt_path <path-to-model> --image_path_or_dir <path-to-image>
+python tools/safety_checker/safety_checker_2.py --ckpt_path <path-to-model> --image_path_or_dir <path-to-image>
+```
+- PyTorch backend
+```
+python tools/safety_checker/safety_checker_1.py --backend pt --image_path_or_dir <path-to-image>
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/docs/en/safety_check.md:18]: Can we combine safety_checker_1.py and safety_checker_1.py into one script like safety_checker.py? And use an argument --version to select which one we would like to use

File: examples/stable_diffusion_v2/tools/safety_checker/safety_checker_1.py (hunk 4 above)
Comment (xijia-tao) [reply examples/stable_diffusion_v2/tools/safety_checker/safety_checker_1.py:207]: Both `safety_setting_1.yaml` and `safety_setting_2.yaml` (they contain nsfw terms) are for safety checker v1. If we do this, we still need `--config` arg for v2. And I think it's good to have safety settings and model settings decoupled.

So maybe we can leave these two arguments unchanged?

**Review by wtomin — COMMENTED** [review]
Can we remove l14_nsfw.ckpt and l14_nsfw.pth? Although they are small models' weights, they still make the source code a bit larger (about 1MB). I think it's better to embed the download links into the code to allow automatic downloading.

File: examples/stable_diffusion_v2/docs/en/safety_check.md (hunk 5 above)
Comment (xijia-tao) [reply examples/stable_diffusion_v2/docs/en/safety_check.md:20]: done

File: examples/stable_diffusion_v2/tools/safety_checker/l14_nsfw.ckpt
Comment (xijia-tao) [reply examples/stable_diffusion_v2/tools/safety_checker/l14_nsfw.ckpt:1]: done

File: examples/stable_diffusion_v2/tools/eval/eval_clip_score.py (hunk 3 above)
Comment (xijia-tao) [reply examples/stable_diffusion_v2/tools/eval/eval_clip_score.py]: done

File: examples/stable_diffusion_v2/tools/eval/eval_clip_i_score.py (hunk 2 above)
Comment (xijia-tao) [reply examples/stable_diffusion_v2/tools/eval/eval_clip_i_score.py:76]: ok

File: examples/stable_diffusion_v2/docs/en/safety_check.md
Comment (xijia-tao) [reply examples/stable_diffusion_v2/docs/en/safety_check.md:1]: done

File: examples/stable_diffusion_v2/text_to_image.py (hunk 1 above)
Comment (xijia-tao) [reply examples/stable_diffusion_v2/text_to_image.py]: done

**Comment by DongyaoZhu** [comment]
I'd try to divide 40 files into smaller groups across several prs

**Comment by xijia-tao** [comment]
> I'd try to divide 40 files into smaller groups across several prs

maybe I do 4 prs for
1. move clip implementation from eval to tools/_common
2. move clip score and fid from eval to tools/eval
3. new codes and doc for safety checker
4. modify text_to_image.py to add safety checker?

**Comment by xijia-tao** [comment]
> > > I'd try to divide 40 files into smaller groups across several prs
> > 
> > 
> > @DongyaoZhu maybe I do 4 prs for
> > 
> > 1. move clip implementation from eval to tools/_common
> > 2. move clip score and fid from eval to tools/eval
> > 3. new codes and doc for safety checker
> > 4. modify text_to_image.py to add safety checker?
> 
> [1, 2] + [3, 4] will do

ok

**Pull request closed without merging**